app/gui/src/providers/backends.ts

```diff
@@ -0,0 +1,100 @@
+import { isCloudCategory, type Category } from '#/layouts/CategorySwitcher/Category'
+import { BackendType } from '#/services/Backend'
+import LocalBackend from '#/services/LocalBackend'
+import ProjectManager, {
+  ProjectManagerEvents,
+  Path as ProjectManagerPath,
+} from '#/services/ProjectManager'
+import RemoteBackend from '#/services/RemoteBackend'
+import HttpClient from '#/utilities/HttpClient'
+import { createContextStore } from '@/providers'
+import { GuiConfig } from '@/providers/guiConfig'
+import { ToValue } from '@/util/reactivity'
+import * as common from 'enso-common'
+import invariant from 'tiny-invariant'
+import { computed, markRaw, onScopeDispose, readonly, ref, toValue } from 'vue'
+import { GetText } from './text'
+
+function useBackends(
+  httpClient: HttpClient,
+  config: ToValue<GuiConfig>,
+  rootDirPath: ToValue<string | undefined>,
+  getText: GetText,
+) {
+  const projectManager = computed(() => {
+    const rootPath = toValue(rootDirPath)
+    if (rootPath == null) return
+    const cfg = toValue(config)
+    if (cfg.projectManagerUrl == null) return
+    const rootDirectory = ProjectManagerPath(rootPath)
+    if (rootDirPath && cfg.projectManagerUrl != null)
+      return new ProjectManager(cfg.projectManagerUrl, rootDirectory)
+    else return null
+  })
+  const localBackend = computed(() =>
+    projectManager.value ? markRaw(new LocalBackend(projectManager.value)) : null,
```

Comment:
Instead of having to remember to `markRaw` every instance, it's better to take care of that at the class level. You can add `markRaw(LocalBackend.prototype)` in the class file.